File: moat-main/dashboard/backend/app/schemas/invention.py
from datetime import datetime
from pydantic import BaseModel, Field

from app.models.invention import InventionDocumentType, InventionStatus


class InventionCreate(BaseModel):
    workspace_id: str | None = None
    matter_id: str | None = None
    title: str = Field(..., max_length=255)
    description: str | None = None
    status: InventionStatus = InventionStatus.draft


class InventionUpdate(BaseModel):
    workspace_id: str | None = None
    matter_id: str | None = None
    title: str | None = Field(None, max_length=255)
    description: str | None = None
    status: InventionStatus | None = None


class InventionDocumentResponse(BaseModel):
    id: str
    invention_id: str
    file_name: str
    file_type: InventionDocumentType
    content_type: str | None = None
    storage_url: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class InventionAnalysisResponse(BaseModel):
    id: str
    invention_id: str
    technical_summary: str | None = None
    innovation_summary: str | None = None
    key_components: list = []
    technical_domains: list = []
    differentiators: list = []
    workflows: list = []
    technical_architecture: dict | None = None
    innovation_highlights: list = []
    confidence_score: float = 0.0
    model_name: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class InventionResponse(BaseModel):
    id: str
    workspace_id: str | None = None
    matter_id: str | None = None
    title: str
    description: str | None = None
    status: InventionStatus
    created_by: str
    created_at: datetime
    updated_at: datetime
    documents: list[InventionDocumentResponse] = []
    latest_analysis: InventionAnalysisResponse | None = None

    class Config:
        from_attributes = True


class InventionAnalyzeRequest(BaseModel):
    force: bool = False


class InventionAnalyzeResponse(BaseModel):
    invention: InventionResponse
    analysis: InventionAnalysisResponse
    task_id: str | None = None

File: moat-main/dashboard/frontend/src/types/invention.ts
export type InventionStatus = "draft" | "ready" | "analyzing" | "analyzed" | "failed" | "archived";
export type InventionDocumentType = "pdf" | "patent" | "technical_document" | "image" | "sketch" | "diagram" | "other";

export interface InventionDocument {
  id: string;
  invention_id: string;
  file_name: string;
  file_type: InventionDocumentType;
  content_type?: string | null;
  storage_url?: string | null;
  created_at: string;
}

export interface InventionAnalysis {
  id: string;
  invention_id: string;
  technical_summary?: string | null;
  innovation_summary?: string | null;
  key_components: Array<Record<string, any>>;
  technical_domains: Array<Record<string, any>>;
  differentiators: Array<Record<string, any>>;
  workflows: Array<Record<string, any>>;
  technical_architecture?: Record<string, any> | null;
  innovation_highlights: Array<Record<string, any>>;
  confidence_score: number;
  model_name?: string | null;
  created_at: string;
}

export interface Invention {
  id: string;
  workspace_id?: string | null;
  matter_id?: string | null;
  title: string;
  description?: string | null;
  status: InventionStatus;
  created_by: string;
  created_at: string;
  updated_at: string;
  documents: InventionDocument[];
  latest_analysis?: InventionAnalysis | null;
}

export interface InventionCreateInput {
  workspace_id?: string | null;
  matter_id?: string | null;
  title: string;
  description?: string | null;
  status?: InventionStatus;
}
